File: src/main/storage/repos/saved-queries.ts
import { and, desc, eq } from 'drizzle-orm';
import {
  type ConsoleBody,
  consoleBodySchema,
  type FilterBody,
  filterBodySchema,
  type SavedConsoleQuery,
  type SavedFilterQuery,
  type SavedQuery,
  savedQuerySchema,
} from '../../../shared/domain/queries';
import { log } from '../../log';
import type { KiraDb } from '../db';
import { savedQueries } from '../schema/saved-queries';

interface SavedQueryRow {
  id: string;
  connectionId: string;
  path: string;
  name: string;
  kind: string;
  body: string;
  pinned: boolean;
  createdAt: string;
  usedAt: string | null;
}

function parseSavedQueryRow(row: SavedQueryRow): SavedQuery | null {
  let body: unknown;
  try {
    body = JSON.parse(row.body);
  } catch {
    log('warn', 'storage/saved-queries', `dropping saved query ${row.id}: body is not valid JSON`);
    return null;
  }
  const parsed = savedQuerySchema.safeParse({ ...row, body });
  if (!parsed.success) {
    log(
      'warn',
      'storage/saved-queries',
      `dropping unparseable saved query ${row.id}: ${parsed.error.message}`,
    );
    return null;
  }
  return parsed.data;
}

async function listByKind<K extends SavedQuery['kind']>(
  db: KiraDb,
  connectionId: string,
  path: string,
  kind: K,
): Promise<Extract<SavedQuery, { kind: K }>[]> {
  const rows = await db
    .select()
    .from(savedQueries)
    .where(
      and(
        eq(savedQueries.connectionId, connectionId),
        eq(savedQueries.path, path),
        eq(savedQueries.kind, kind),
      ),
    )
    .orderBy(desc(savedQueries.pinned), desc(savedQueries.usedAt), savedQueries.name);
  const out: Extract<SavedQuery, { kind: K }>[] = [];
  for (const row of rows) {
    const parsed = parseSavedQueryRow(row);
    if (parsed && parsed.kind === kind) out.push(parsed as Extract<SavedQuery, { kind: K }>);
  }
  return out;
}

export async function listSavedFilters(
  db: KiraDb,
  connectionId: string,
  path: string,
): Promise<SavedFilterQuery[]> {
  return listByKind(db, connectionId, path, 'filter');
}

export async function listSavedConsoleQueries(
  db: KiraDb,
  connectionId: string,
  path: string,
): Promise<SavedConsoleQuery[]> {
  return listByKind(db, connectionId, path, 'console');
}

async function insertSavedQuery(
  db: KiraDb,
  input: {
    connectionId: string;
    path: string;
    name: string;
    kind: SavedQuery['kind'];
    body: string;
    pinned: boolean;
  },
): Promise<SavedQuery> {
  const id = crypto.randomUUID();
  const now = new Date().toISOString();
  await db.insert(savedQueries).values({
    id,
    connectionId: input.connectionId,
    path: input.path,
    name: input.name,
    kind: input.kind,
    body: input.body,
    pinned: input.pinned,
    createdAt: now,
    usedAt: now,
  });
  const row = (await db.select().from(savedQueries).where(eq(savedQueries.id, id)))[0];
  const parsed = row ? parseSavedQueryRow(row) : null;
  if (!parsed) throw new Error('insertSavedQuery: freshly inserted row failed to parse');
  return parsed;
}

export async function saveFilter(
  db: KiraDb,
  input: { connectionId: string; path: string; name: string; body: FilterBody; pinned: boolean },
): Promise<SavedFilterQuery> {
  const saved = await insertSavedQuery(db, {
    ...input,
    kind: 'filter',
    body: JSON.stringify(filterBodySchema.parse(input.body)),
  });
  if (saved.kind !== 'filter') throw new Error('saveFilter: inserted row is not kind filter');
  return saved;
}

export async function saveConsoleQuery(
  db: KiraDb,
  input: { connectionId: string; path: string; name: string; body: ConsoleBody; pinned: boolean },
): Promise<SavedConsoleQuery> {
  const saved = await insertSavedQuery(db, {
    ...input,
    kind: 'console',
    body: JSON.stringify(consoleBodySchema.parse(input.body)),
  });
  if (saved.kind !== 'console') {
    throw new Error('saveConsoleQuery: inserted row is not kind console');
  }
  return saved;
}

// Kind-agnostic: `id` alone identifies the row, and neither op touches `body`/`kind` — one
// implementation serves both saved filters (§8.5) and saved console queries (§8.14).
export async function updateSavedQuery(
  db: KiraDb,
  id: string,
  patch: { name?: string; pinned?: boolean },
): Promise<SavedQuery> {
  await db
    .update(savedQueries)
    .set({
      ...(patch.name !== undefined ? { name: patch.name } : {}),
      ...(patch.pinned !== undefined ? { pinned: patch.pinned } : {}),
    })
    .where(eq(savedQueries.id, id));
  const row = (await db.select().from(savedQueries).where(eq(savedQueries.id, id)))[0];
  const parsed = row ? parseSavedQueryRow(row) : null;
  if (!parsed) throw new Error(`updateSavedQuery: row ${id} not found after update`);
  return parsed;
}

export async function deleteSavedQuery(db: KiraDb, id: string): Promise<void> {
  await db.delete(savedQueries).where(eq(savedQueries.id, id));
}

export async function touchSavedQuery(db: KiraDb, id: string): Promise<void> {
  await db
    .update(savedQueries)
    .set({ usedAt: new Date().toISOString() })
    .where(eq(savedQueries.id, id));
}

File: apps/kira-studio/internal/storage/repos/saved_queries.go
package repos

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/id"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

const savedQuerySelectColumns = `id, connection_id, path, name, kind, body, pinned, created_at, used_at`

type SavedQueriesRepo struct {
	DB *sql.DB
}

// scanSavedQueryRow scans one row and validates it the same way saved-queries.ts's
// parseSavedQueryRow does: an unparseable body, or one that doesn't match its own kind's shape,
// drops the row (nil, nil) rather than propagating it.
func scanSavedQueryRow(row rowScanner) (*model.SavedQuery, error) {
	var (
		q      model.SavedQuery
		pinned int
		usedAt sql.NullString
		body   string
	)
	if err := row.Scan(&q.ID, &q.ConnectionID, &q.Path, &q.Name, &q.Kind, &body, &pinned, &q.CreatedAt, &usedAt); err != nil {
		return nil, err
	}
	q.Pinned = pinned != 0
	if usedAt.Valid {
		q.UsedAt = &usedAt.String
	}

	if !json.Valid([]byte(body)) {
		slog.Warn("dropping saved query: body is not valid JSON", "scope", "storage/saved-queries", "id", q.ID)
		return nil, nil
	}
	switch q.Kind {
	case "filter":
		var fb model.FilterBody
		// DisallowUnknownFields is what actually distinguishes a filter body from a console
		// body: FilterBody/ConsoleBody's fields don't overlap, and json.Unmarshal alone
		// silently ignores a console body's "text" key when decoding into FilterBody instead
		// of rejecting it.
		dec := json.NewDecoder(strings.NewReader(body))
		dec.DisallowUnknownFields()
		if err := dec.Decode(&fb); err != nil {
			slog.Warn("dropping saved query: body does not match filter shape", "scope", "storage/saved-queries", "id", q.ID)
			return nil, nil
		}
	case "console":
		var cb model.ConsoleBody
		dec := json.NewDecoder(strings.NewReader(body))
		dec.DisallowUnknownFields()
		if err := dec.Decode(&cb); err != nil {
			slog.Warn("dropping saved query: body does not match console shape", "scope", "storage/saved-queries", "id", q.ID)
			return nil, nil
		}
	default:
		slog.Warn("dropping saved query: unrecognised kind", "scope", "storage/saved-queries", "id", q.ID, "kind", q.Kind)
		return nil, nil
	}
	q.Body = json.RawMessage(body)
	return &q, nil
}

func (r *SavedQueriesRepo) listByKind(connectionID, path, kind string) ([]model.SavedQuery, error) {
	rows, err := r.DB.Query(
		`SELECT `+savedQuerySelectColumns+`
		   FROM saved_queries
		  WHERE connection_id = ? AND path = ? AND kind = ?
		  ORDER BY pinned DESC, used_at DESC, name`,
		connectionID, path, kind,
	)
	if err != nil {
		return nil, fmt.Errorf("repos/saved_queries: query: %w", err)
	}
	defer rows.Close()

	out := []model.SavedQuery{}
	for rows.Next() {
		q, err := scanSavedQueryRow(rows)
		if err != nil {
			return nil, fmt.Errorf("repos/saved_queries: scan: %w", err)
		}
		if q != nil {
			out = append(out, *q)
		}
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("repos/saved_queries: rows: %w", err)
	}
	return out, nil
}

func (r *SavedQueriesRepo) ListFilters(connectionID, path string) ([]model.SavedQuery, error) {
	return r.listByKind(connectionID, path, "filter")
}

func (r *SavedQueriesRepo) ListConsole(connectionID, path string) ([]model.SavedQuery, error) {
	return r.listByKind(connectionID, path, "console")
}

func (r *SavedQueriesRepo) insert(connectionID, path, name, kind string, body []byte) (model.SavedQuery, error) {
	if err := model.ValidSavedQueryName(name); err != nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: %w", err)
	}
	newID := id.New()
	now := model.NowISO()
	if _, err := r.DB.Exec(
		`INSERT INTO saved_queries (id, connection_id, path, name, kind, body, pinned, created_at, used_at)
		 VALUES (?, ?, ?, ?, ?, ?, 0, ?, ?)`,
		newID, connectionID, path, name, kind, string(body), now, now,
	); err != nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: insert: %w", err)
	}
	return r.mustGet(newID)
}

func (r *SavedQueriesRepo) mustGet(queryID string) (model.SavedQuery, error) {
	row := r.DB.QueryRow(`SELECT `+savedQuerySelectColumns+` FROM saved_queries WHERE id = ?`, queryID)
	q, err := scanSavedQueryRow(row)
	if err != nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: get %s: %w", queryID, err)
	}
	if q == nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: row %s not found or failed to parse", queryID)
	}
	return *q, nil
}

func (r *SavedQueriesRepo) SaveFilter(connectionID, path, name string, body model.FilterBody, pinned bool) (model.SavedQuery, error) {
	encoded, err := json.Marshal(body)
	if err != nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: encode filter body: %w", err)
	}
	q, err := r.insert(connectionID, path, name, "filter", encoded)
	if err != nil {
		return model.SavedQuery{}, err
	}
	if pinned {
		if err := r.setPinned(q.ID, true); err != nil {
			return model.SavedQuery{}, err
		}
		return r.mustGet(q.ID)
	}
	return q, nil
}

func (r *SavedQueriesRepo) SaveConsole(connectionID, path, name string, body model.ConsoleBody, pinned bool) (model.SavedQuery, error) {
	encoded, err := json.Marshal(body)
	if err != nil {
		return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: encode console body: %w", err)
	}
	q, err := r.insert(connectionID, path, name, "console", encoded)
	if err != nil {
		return model.SavedQuery{}, err
	}
	if pinned {
		if err := r.setPinned(q.ID, true); err != nil {
			return model.SavedQuery{}, err
		}
		return r.mustGet(q.ID)
	}
	return q, nil
}

func (r *SavedQueriesRepo) setPinned(queryID string, pinned bool) error {
	if _, err := r.DB.Exec(`UPDATE saved_queries SET pinned = ? WHERE id = ?`, boolToInt(pinned), queryID); err != nil {
		return fmt.Errorf("repos/saved_queries: set pinned %s: %w", queryID, err)
	}
	return nil
}

// Update is kind-agnostic (queries.ts's own updateSavedQuery): id alone identifies the row, and
// neither field touches body/kind.
func (r *SavedQueriesRepo) Update(queryID string, patch model.SavedQueryPatch) (model.SavedQuery, error) {
	if patch.Name != nil {
		if err := model.ValidSavedQueryName(*patch.Name); err != nil {
			return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: %w", err)
		}
		if _, err := r.DB.Exec(`UPDATE saved_queries SET name = ? WHERE id = ?`, *patch.Name, queryID); err != nil {
			return model.SavedQuery{}, fmt.Errorf("repos/saved_queries: update name %s: %w", queryID, err)
		}
	}
	if patch.Pinned != nil {
		if err := r.setPinned(queryID, *patch.Pinned); err != nil {
			return model.SavedQuery{}, err
		}
	}
	return r.mustGet(queryID)
}

func (r *SavedQueriesRepo) Delete(queryID string) error {
	if _, err := r.DB.Exec(`DELETE FROM saved_queries WHERE id = ?`, queryID); err != nil {
		return fmt.Errorf("repos/saved_queries: delete %s: %w", queryID, err)
	}
	return nil
}

func (r *SavedQueriesRepo) Touch(queryID string) error {
	if _, err := r.DB.Exec(`UPDATE saved_queries SET used_at = ? WHERE id = ?`, model.NowISO(), queryID); err != nil {
		return fmt.Errorf("repos/saved_queries: touch %s: %w", queryID, err)
	}
	return nil
}
